### ISSUE TITLE
BUG: Cannot parse formula with a function

### ISSUE THREAD

**Issue by bashtage** [matthewwardrop/formulaic#28]
I can't seel to get functions in formulas to work.  This is a pretty basic one I think.

```python
import numpy as np
from pandas import DataFrame, Categorical

formula = "y ~ 1 +np.exp(x1)"
y = np.random.randn(1000)
x1 = np.random.randn(1000)
d = np.random.randint(0, 4, 1000)
d = Categorical(d)
data = DataFrame({"y": y, "x1": x1, "d": d})
data["Intercept"] = 1.0
model_matrix(formula, data,context=0)
```
produces

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-19-930162752fbf> in <module>
      9 data = DataFrame({"y": y, "x1": x1, "d": d})
     10 data["Intercept"] = 1.0
---> 11 model_matrix(formula, data,context=0)

c:\git\formulaic\formulaic\sugar.py in model_matrix(formula, data, context, **kwargs)
     12         else:
     13             context = None  # pragma: no cover
---> 14     return Formula(formula).get_model_matrix(data, context=context, **kwargs)

c:\git\formulaic\formulaic\formula.py in get_model_matrix(self, data, context, materializer, ensure_full_rank, **kwargs)
     61         if not inspect.isclass(materializer) or not issubclass(materializer, FormulaMaterializer):
     62             raise FormulaMaterializerInvalidError("Materializers must be subclasses of `formulaic.materializers.FormulaMaterializer`.")
---> 63         return materializer(data, context=context or {}).get_model_matrix(self, ensure_full_rank=ensure_full_rank, **kwargs)
     64
     65     def differentiate(self, *vars, use_sympy=False):

c:\git\formulaic\formulaic\materializers\base.py in get_model_matrix(self, spec, ensure_full_rank, na_action, output)
    108         # Step 0: Check whether formula separators are in play, and if so, recurse.
    109         if isinstance(spec.formula.terms, tuple):
--> 110             return tuple(
    111                 self.get_model_matrix(Formula(terms), ensure_full_rank=ensure_full_rank, na_action=na_action, output=output)
    112                 for terms in spec.formula.terms

c:\git\formulaic\formulaic\materializers\base.py in <genexpr>(.0)
    109         if isinstance(spec.formula.terms, tuple):
    110             return tuple(
--> 111                 self.get_model_matrix(Formula(terms), ensure_full_rank=ensure_full_rank, na_action=na_action, output=output)
    112                 for terms in spec.formula.terms
    113             )

c:\git\formulaic\formulaic\materializers\base.py in get_model_matrix(self, spec, ensure_full_rank, na_action, output)
    118         for term in spec.formula.terms:
    119             for factor in term.factors:
--> 120                 self._evaluate_factor(factor, spec, drop_rows)
    121
    122         drop_rows = sorted(drop_rows)

c:\git\formulaic\formulaic\materializers\base.py in _evaluate_factor(self, factor, spec, drop_rows)
    269                 value = self._lookup(factor.expr)
    270             elif factor.eval_method.value == 'python':
--> 271                 value = self._evaluate(factor.expr, factor.metadata, spec)
    272             elif factor.eval_method.value == 'literal':
    273                 value = EvaluatedFactor(factor, self._evaluate(factor.expr, factor.metadata, spec), kind='constant')

c:\git\formulaic\formulaic\materializers\base.py in _evaluate(self, expr, metadata, spec)
    306
    307     def _evaluate(self, expr, metadata, spec):
--> 308         return stateful_eval(expr, self.layered_context, {expr: metadata}, spec.transform_state, spec)
    309
    310     def _is_categorical(self, values):

c:\git\formulaic\formulaic\utils\stateful_transforms.py in stateful_eval(expr, env, metadata, state, spec)
     66     stateful_nodes = {}
     67     for node in ast.walk(code):
---> 68         if isinstance(node, ast.Call) and getattr(env.get(node.func.id), '__is_stateful_transform__', False):
     69             stateful_nodes[astor.to_source(node).strip()] = node
     70

AttributeError: 'Attribute' object has no attribute 'id'
```

**Comment by matthewwardrop** [comment]
Thanks so much for reporting this, and for the patch.

I dug a bit deeper into this, and found out what's going on. When I tested functions, I only tested functions which were directly available in the current namespace (either by importing or direct definition). It turns out the abstract syntax tree node for a function accessed indirectly is different (an Attribute vs. a Name), which I hadn't properly appreciated. I put in the more general fix :).
